### PR TITLE
feat(storage): Store blob content-type when using FSBackend

### PR DESCRIPTION
**Pull request by jpetazzo**
I'm not super excited about this implementation because it
feels like I'm peppering the code with `+".type"` everywhere,
and many actions that used to be atomic (like moving a file)
aren't anymore. But perhaps that's an acceptable trade-off.

This is an alternate implementation to #115.

### CODE REVIEW

**Comment by flokli** [comment]
I'd personally prefer using xattrs for per-file metadata. It's feels much cleaner for "metadata for a file", really is just "key value pairs", doesn't litter text files with metadata next to the real files, and when you (re)move the file, things are atomic.

IIRC, they only exist on Linux and Darwin, and only on some filesystems, but this is probably a bullet we're fine to bite.

**Comment by jpetazzo** [comment]
I was worried about filesystem support; but if you think it's a good tradeoff (especially since there is also the GCS backend support which is independent) I could take a stab at it, for sure.

@tazjin wdyt?

**Comment by tazjin** [comment]
> IIRC, they only exist on Linux and Darwin, and only on some filesystems, but this is probably a bullet we're fine to bite.

In that case, what should the behaviour be on filesystems that don't support it? Error on startup? Return potentially malformed responses?

I think that the solution from the other PR that was merged (ensure Content-Type is set at serving time) is better, but it causes an annoying interface discrepancy (as GCS only lets us set the Content-Type *before* serving).

File: storage/filesystem.go [hunk 1]
```diff
@@ -66,6 +66,18 @@ func (b *FSBackend) Persist(ctx context.Context, key, _type string, f Persister)
 	}
 	defer file.Close()
 
+	typefile, err := os.OpenFile(full+".type", os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
```
Comment (jonjohnsonjr) [inline storage/filesystem.go:69]: What about something slightly more generic like `full+".headers"` to store arbitrary header responses?

This way you could throw `Docker-Content-Digest` in there, as well as any other headers you might need. Seems like a nice flexible escape hatch.

**Comment by flokli** [comment]
> > IIRC, they only exist on Linux and Darwin, and only on some filesystems, but this is probably a bullet we're fine to bite.
> 
> In that case, what should the behaviour be on filesystems that don't support it? Error on startup? Return potentially malformed responses?

The number of file systems not supporting xattrs is super small. I think it's fine to settle on xattrs for now, and in case there's real-world demand, we can still implement a not-so-atomic shim with separate files, which is used as a fallback.

> I think that the solution from the other PR that was merged (ensure Content-Type is set at serving time) is better, but it causes an annoying interface discrepancy (as GCS only lets us set the Content-Type _before_ serving).

We can avoid storing the content type, as we can derive it from the request URL, but the `Docker-Content-Digest` certainly is something we don't want to compute every time.

**Comment by jonjohnsonjr** [comment]
> We can avoid storing the content type, as we can derive it from the request URL

This is only an approximation of the proper registry semantics and won't work if we wanted to do multi-platform images, for example.

**Comment by flokli** [comment]
> > We can avoid storing the content type, as we can derive it from the request URL
> 
> This is only an approximation of the proper registry semantics and won't work if we wanted to do multi-platform images, for example.

Valid point! So we should really switch to annotating content-types as file metadata, and pass these out to users if we have it.

**Comment by tazjin** [comment]
Superseded by #127  :)

**Pull request closed without merging**